sys/doip/doip.c

```diff
@@ -0,0 +1,193 @@
+    for (i = 0; i < dlen; i++) {
+        printf("0x%x ", data[i]);
+    }
+    printf("\n");
+}
+
+static int doip_print_msg_parsed(uint8_t *data, uint32_t dlen)
+{
+    uint16_t payload_type = 0;
+    uint32_t payload_len = 0;
+    uint16_t source = 0;
+    uint16_t target = 0;
+    uint32_t i = 0;
+
+
+    if (dlen < 8 || data == NULL) {     //basic DoIP is always 8 bytes. TA and SA can be dropped
+        return -1;
+    }
+
+    if (data[0] == 0x0) {
+        puts("DoIp version: 01");
+    }
+    if (data[0] == 0x02) {
+        puts("DoIP version: 02");
+    }
+
+    payload_type = (data[2] << 8 | data[3]);
+    //printf("payload_type: %x\n", payload_type);
+    switch (payload_type) {
+        case DOIP_GENERIC_DOIP_HEADER_NEGATIVE_ACK:
+            puts("Payload Type: Generic negative ACK");
+            break;
+        case DOIP_VEHICLE_IDENTIFICATION_REQUEST:
+            puts("Payload Type: Vehicle Identification Request");
+            break;
+        case DOIP_VEHICLE_IDENTIFICATION_REQUEST_WITH_VIN:
+            puts("Payload Type: Vehicle Identification Request with VIN");
+            break;
+        case DOIP_VEHICLE_IDENTIFICATION_REQUEST_WITH_EID:
+            puts("Payload Type: Vehicle Identification Request with EID");
+            break;
+        case DOIP_VEHICLE_IDENTIFCATION_RESPONSE:
+            puts("Payload Type: Vehicle Identification Reply");
+            break;
+        case DOIP_DIAGNOSTIC_MESSAGE:
+            puts("Payload Type: Diagnostic Message Request");
+            break;
+        case DOIP_DIAGNOSTIC_MESSAGE_POSITIVE_ACK:
+            puts("Payload Type: Diagnostic Message Positive Acknowledge");
+            break;
+
+        //TODO: implement others!
+        default:
+            puts("Unknown payload type");
+            break;
+    }
+    payload_len = (data[4] << 24 | data[5] << 16 | data[6] << 8 | data[7]);
+    printf("Payload length: %ld\n", payload_len);
+
+    if (dlen < 12) {
+        return 0;
+    }
+    source = (data[8] << 8 | data[9]);
+    target = (data[10] << 8 | data[11]);
+    printf("Source address: %x, target address: %x\n", source, target);
+
+    puts("UDS data:");
+    for (i = 12; i < dlen; i++) {
+        printf("0x%x ", data[i]);
+    }
+    printf("\n");
+
+
+    return 0;
+
+
+}
+
+int doip_send_udp(doip_sa sa, doip_ta ta, uint16_t payload_type, uint8_t *data,
+                  uint32_t dlen, char *ip_addr)
+{
+    int ret = 0;
+    sock_udp_ep_t remote = SOCK_IPV4_EP_ANY;
+    sock_udp_ep_t local = SOCK_IPV4_EP_ANY;
+    sock_udp_t sock;
+
+    local.port = 41123;
+
+    uint32_t payload_len = dlen + 4;        //TA and SA should also be included in the payload length!
+    //Not sure if it's always true though... What if TA and SA are 0 (e.g. with Veh. Id. Req)? Then those aren't added to payload, thus, no +4?
+
+    int msglen = 0;         //Total length of message to be sent over UDP
+
+    uint8_t buf[50] = { '\0' };
+    if (sock_udp_create(&sock, &local, NULL, 0)) {
```

Comment:
move this socket creation to a new `sock_doip_create`function

Reply:
A doip socket can be used so the underlying udp socket doesn't have to be created and closed every time you want to send something